### DISCUSSION TITLE
I'm curious about the ADC

### DISCUSSION THREAD

**Discussion by herbkonik** [dbuezas/lgt8fx#251] [General]
Hi,
after building a gauge to meassure and store current i found some things at the ADC where i'm curious about.
First: the max value i get is 4064 (board powerd at 5V Pin, ADC pin at same source)
Secound: I get a different output at some ADC Pins if i put the same voltage at it. E.g. at one pin i get 1500, with the same voltage i get 1540 at some different pins but others show also 1500.

I try'd a few of the purple boards now und each shows the same behavior, - expecting a little different at the difference.

**Comment by LaZsolt** [comment]
> the max value i get is 4064

The reason is, the ADC has a gain error. The ADC internal value reach the maximum value (4095) sooner than the input reach the reference voltage. The ADC can not masure the voltage slightly smaller than the reference voltage. So the last 31 steps lost. This behaviour compensated by software so this is the reason for the maximum value of 4064.

> at one pin i get 1500, with the same voltage i get 1540 at some different pins

The difference between 1500 and 1540 is around 1%. I think it is acceptable because no electrical parameters were written in the databook.

By the way. Did you disabled pull up resistors and digital input functions during measure?
In SSOP20 package some ADC input pins have internally connected to other ports. In this case in the internally connected other port must be disabled pull up resistors and digital input function also. (In SSOP20 package inputs ADC6 and ADC7.)

**Comment by herbkonik** [comment]
Thank you
The maximum value of 4064 isn't a real prob, if one know it's easy to compensate... 

The curious thing with that difference is to me: there are only two values, nothing between, only one or the other.
One idea: dosn't the 328P have two ADCs ? May be that's the reason, the different comes from the different ADCs... I'l take a look closer.
If so, I'l find a workaround, e.g. using only one ov the ADCs.
The gauge is build to make long time comparisation of solar panels, for that 1% differece is to much.

The Boards i'm using are with the LQFP32 package, - is there a similar situation with double connections ?

**Reply by LaZsolt** [reply]
>The Boards i'm using are with the LQFP32 package, - is there a similar situation with double connections ?

In LQPF32 package no double connections with any ADC inputs.

**Comment by dbuezas** [comment]
> The gauge is build to make long time comparisation of solar panels, for that 1% differece is to much.

Maybe you can use the differential amplifier for that.

BTW, I remember there's a register to turn off the digital input circuitry to reduce ADC noise, maybe you can try that too

**Comment by jayzakk** [comment]
Just to sum it up:

We're not talking about accuracy, but different readings of the same voltage depending on the input pin used?

I doubt the LGTF is less accurate at the Atmega, which has 10 bit resolution with +-2 LSB, makes about 0,4%. As +-2 LSB is quite some regular value on that cheap ADCs, and LGTF has 12 Bits, it must be much better.

Different pins reading different values, that's really interesting. Would you mind to tell us if there are groups of pins maybe, or which ones are off by large? Asides from pulling resistors, it could be the multiplexer, or even different types of internal circuitry (which has really good documentation in the datasheet).

Given that you drive the inputs strong enough (guess you had to build up a resistor divider if measuring solar cell voltage).

I didn't see such thing until now, but have to admit that I never compared all the inputs....

**Comment by herbkonik** [comment]
Yes, we're not talking about accuracy, but different readings of the same voltage depending on the input pin used.
In my mind  the accuracy of the LGT is some better, i never got a stable output like at the LGT at an Atmega. Where the Atmega output always shows a jumping of it's +-2 LSB, the LGT output jumps only 1 ( not +- ).
Of course same hardware around and same Sketch, - Sketch expecting the setup to 12 Bit resolution.

`Would you mind to tell us...`
Of course... 
Like i wrote above, i want to try something into this behavior, also how much the difference is at different input voltage.

I don't measure the voltage, only the (short sircuit) current, so no  resistor divider is involved. :)  
For the Current sensor i use a LM358P as differential amplifier, i noticed this "different readings of the same voltage" as i made the calibration of the sensor.

@dbuezas : turning at the registers is much over my skills... ;)

**Reply by dbuezas** [reply]
> turning at the registers is much over my skills... ;)

You can definitely do it! :)
![Screenshot_20230225-101810](https://user-images.githubusercontent.com/777196/221349264-92fdd8f6-461a-4c58-99dd-fd7851156f68.png)

For example:
```cpp
bitSet(DIDR0, PC0D);  // disable digital input (reduce noise) A0
  // bitSet(DIDR0, PC1D);  // disable digital input (reduce noise) A1
  // bitSet(DIDR0, PC2D);  // disable digital input (reduce noise) A2
  // bitSet(DIDR0, PC3D);  // disable digital input (reduce noise) A3
```

**Reply by vklimk** [reply]
> @dbuezas : turning at the registers is much over my skills... ;)

Try to disable pull-up resistors globally using this line of code:
MCUCR = (1 << PUD);

**Reply by dbuezas** [reply]
I think you need to flip WCE first, and then in a second statement set the PUD bit:
```cpp
MCUCR |= 1<<WCE;
MCUCR |= 1<<PUD;
```

**Reply by LaZsolt** [reply]
Better to use this code:
```
noInterrupts();
uint8_t temp = MCUCR | 1<<PUD;
MCUCR = 1<<MWCE;
MCUCR = temp;
interrupts();
```
Because WCE is a 7th bit, MWCE is a 0th bit.

**Reply by LaZsolt** [reply]
Three people needed to disable pull-up resistors. (:

**Reply by dbuezas** [reply]
Is this any different than disabling just not enabling the pullups on specific pins?
I wonder if disabling the digital circuitry also disables pullups

**Reply by LaZsolt** [reply]
According to the ATmega328/P databook:

PUD Pull-up Disable bit:
When this bit is written to one, the pull ups in the I/O ports are disabled even if the DDxn and PORTxn registers are configured to enable the pull ups ({DDxn, PORTxn} = 0b01).

> I wonder if disabling the digital circuitry also disables pullups

Not tested atm.

**Comment by dbuezas** [comment]
If you need to measure voltage differences between two pins, the lgt328p has an internal differential amplifier with configurable gain, see here: https://github.com/dbuezas/lgt8fx/blob/master/docs/differential-amplifier/readme.md

Y in theory you can measure differences of 5v/2^17= 38uV.

**Reply by dbuezas** [reply]
The 17 comes from the 12 bits of adc resolution plus the 32x gain

**Comment by herbkonik** [comment]
A little update to this storry.
First: The reason for the different voltage between two pins is NOT a ADC problem. It seams that the (board) layout of my gauge is the base of it.
To check the ADC i dont use any Bob, but put the cables directly to the pins. This shows at each pin exactly the same output ! 
Exactly means: Dos not show only one LSB over al 8 Pins (AD 9 i didn't test).
One other nice result: The output also not differs between different boards ( i tested 4 of the purple, all i have atm):

But what the test also shows: The ADC don't work up to 5V, it returns the max value of 4064 already at 4.65 V. Because i tested "board powerd by USB" i checked the voltage of the USB port, its stable by 5.06 V.

**Reply by dbuezas** [reply]
If you power it through usb, the mcu is behind a diode. The 4.65v is 5v minus the diode drop.
You can either power it via the 5v pin or just short the the diode right by the micro usb port.
AFAIK the diode just makes sure no power flows towards the usb from the vin/5v pins. It also acts as a fuse if you short the rails while connected to usb (ask me how I found that out hehe).

**Reply by herbkonik** [reply]
Ahh, that explanes it...
At regular use i always power the boards by the 5 V pin, only for tests, where i need the serial monitor, i use the USB

**Reply by lalo-uy** [reply]
If the CPU is powered at 4.65 then the ADC should show 4093 at 4.65V input.
The ADC si absolute or ratiometric as the Atmega328 ?

**Reply by dbuezas** [reply]
I think it is relative to AVCC by default (the + rail), but you can use other internal or external reference voltages.

I played with that [in the web oscilloscope project](https://github.com/dbuezas/arduino-web-oscilloscope/blob/master/arduino/src/adc.h#L40)

BTW, there are also two internal resistive dividers (1/5 and 4/5) controlled via the ADMUX register. I used them [here](https://github.com/dbuezas/arduino-web-oscilloscope/blob/master/arduino/src/adc.h#L90) too

**Reply by LaZsolt** [reply]
@lalo-uy 
The ACD of LG8F328P has a big gain error. Because of this the ADC reach its maximum before the input reach the reference voltage.  This behavior is compensated by software, so analogRead() return value never exceed 4064.

**Comment by gadefox** [comment]
I'm stuck with this:
```
DIDR0 &= ~(1 << PC0D); // enable digital input A0
ADCSRD = 1 << VDS0; // ADC0
ADCSRD |= 1 << REFS2;
VCAL = VCAL3; // 4V096 ref.
ADMUX = 13; // IVREF

ADCSRA |= (1 << ADSC);
while (ADCSRA & (1 << ADSC));
return ADC;
```
`IVREF` means `4.096V` right? The input voltage is ~ 1.8V, but I'm getting 4095. Any idea??

EDIT: the same with lower voltage e.g. 720mV:
VCC: 
 4V096: 
  1V5: avg=948 min=946 max=950
  4V5: avg=3797 min=3792 max=3803
A0: 
 4V096: 
  1V5: avg=144 min=144 max=145 -> **144*5=720mV**
  4V5: avg=584 min=584 max=585 -> **584*1.25=730mV**
  IVREF: avg=4095 min=4095 max=4095 -> **??**

**Reply by dbuezas** [reply]
Ref voltages can't be higher than vin. Maybe try the 1v ref and the 1/5 voltage divider
